python/whylogs/core/model_performance_metrics/confusion_matrix.py

```diff
@@ -0,0 +1,244 @@
+                    " confusion matrix will be larger than is recommended with whylogs current"
+                    " representation of the model metric for a confusion matrix of this size."
+                )
+            if labels_size > MODEL_METRICS_MAX_LABELS:
+                raise ValueError(
+                    f"The initialized confusion matrix has {labels_size} labels and the resulting"
+                    " confusion matrix will be larger than is supported by whylogs current"
+                    " representation of the model metric for a confusion matrix of this size,"
+                    " selectively log the most important labels or configure the threshold of"
+                    " {MODEL_METRICS_MAX_LABELS} higher by setting MODEL_METRICS_MAX_LABELS."
+                )
+
+            self.labels = sorted(labels)
+        else:
+            self.labels = list()
+
+        self.confusion_matrix: Dict[Tuple[int, int], DistributionMetric] = dict()
+        self.default_config = MetricConfig()
+
+    def add(
+        self,
+        predictions: List[Union[str, int, bool, float]],
+        targets: List[Union[str, int, bool, float]],
+        scores: Optional[List[float]],
+    ):
+        """
+        Function adds predictions and targets to confusion matrix with scores.
+
+        Args:
+            predictions (List[Union[str, int, bool]]):
+            targets (List[Union[str, int, bool]]):
+            scores (List[float]):
+
+        Raises:
+            NotImplementedError: in case targets do not fall into binary or
+            multiclass suport
+            ValueError: incase missing validation or predictions
+        """
+        if not isinstance(targets, list):
+            targets = [targets]
+        if not isinstance(predictions, list):
+            predictions = [predictions]
+
+        if scores is None:
+            scores = [1.0 for _ in range(len(targets))]
+
+        if len(targets) != len(predictions):
+            raise ValueError("both targets and predictions need to have the same length")
+
+        targets_indx = _encode_to_integers(targets, self.labels)
+        prediction_indx = _encode_to_integers(predictions, self.labels)
+
+        for ind in range(len(predictions)):
+            entry_key = prediction_indx[ind], targets_indx[ind]
+            if entry_key not in self.confusion_matrix:
+                self.confusion_matrix[entry_key] = DistributionMetric.zero(self.default_config)
+            data = PreprocessedColumn.apply([scores[ind]])
+            self.confusion_matrix[entry_key].columnar_update(data)
+
+    def merge(self, other_cm):
+        """
+        Merge two seperate confusion matrix which may or may not overlap in labels.
+
+        Args:
+              other_cm (Optional[ConfusionMatrix]): confusion_matrix to merge with self
+        Returns:
+              ConfusionMatrix: merged confusion_matrix
+        """
+        # TODO: always return new objects
```

Comment:
Is this still a todo? Looks like you are returning everywhere, does one just need adjustment?

Reply:
its an older todo, the where the current labels are empty or where the other_cm is None we don't create a new object.